<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calibrated RUV2. — vruv2 • vicar</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Calibrated RUV2. — vruv2" />
<meta property="og:description" content="This function will perform a variant of Removing Unwanted Variation
2-step (RUV2) (Gagnon-Bartsch et al, 2013), where we include a
variance inflation parameter in the factor analysis." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">vicar</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1-11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/customFA.html">Customizable Factor Analyses</a>
    </li>
    <li>
      <a href="../articles/custom_prior.html">Custom Prior for RUVB</a>
    </li>
    <li>
      <a href="../articles/sample_analysis.html">Sample Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/dcgerard/vicar/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calibrated RUV2.</h1>
    <small class="dont-index">Source: <a href='http://github.com/dcgerard/vicar/blob/master/R/ruv2.R'><code>R/ruv2.R</code></a></small>
    <div class="hidden name"><code>vruv2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function will perform a variant of Removing Unwanted Variation
2-step (RUV2) (Gagnon-Bartsch et al, 2013), where we include a
variance inflation parameter in the factor analysis.</p>
    </div>

    <pre class="usage"><span class='fu'>vruv2</span><span class='op'>(</span>
  <span class='va'>Y</span>,
  <span class='va'>X</span>,
  <span class='va'>ctl</span>,
  k <span class='op'>=</span> <span class='cn'>NULL</span>,
  cov_of_interest <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>X</span><span class='op'>)</span>,
  likelihood <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"t"</span>, <span class='st'>"normal"</span><span class='op'>)</span>,
  limmashrink <span class='op'>=</span> <span class='cn'>TRUE</span>,
  degrees_freedom <span class='op'>=</span> <span class='cn'>NULL</span>,
  include_intercept <span class='op'>=</span> <span class='cn'>TRUE</span>,
  gls <span class='op'>=</span> <span class='cn'>TRUE</span>,
  fa_func <span class='op'>=</span> <span class='va'>pca_2step</span>,
  fa_args <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='op'>)</span>,
  use_factor <span class='op'>=</span> <span class='cn'>FALSE</span>,
  force_check <span class='op'>=</span> <span class='cn'>TRUE</span>,
  fa_limmashrink <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>Y</th>
      <td><p>A matrix of numerics. These are the response variables
where each column has its own variance. In a gene expression
study, the rows are the individuals and the columns are the
genes.</p></td>
    </tr>
    <tr>
      <th>X</th>
      <td><p>A matrix of numerics. The covariates of interest.</p></td>
    </tr>
    <tr>
      <th>ctl</th>
      <td><p>A vector of logicals of length <code><a href='https://rdrr.io/r/base/nrow.html'>ncol(Y)</a></code>. If
position i is <code>TRUE</code> then position i is considered a
negative control.</p></td>
    </tr>
    <tr>
      <th>k</th>
      <td><p>A non-negative integer.The number of unobserved
confounders. If not specified and the R package sva is
installed, then this function will estimate the number of
hidden confounders using the methods of Buja and Eyuboglu
(1992).</p></td>
    </tr>
    <tr>
      <th>cov_of_interest</th>
      <td><p>A vector of positive integers. The column
numbers of the covariates in X whose coefficients you are
interested in. The rest are considered nuisance parameters and
are regressed out by OLS.</p></td>
    </tr>
    <tr>
      <th>likelihood</th>
      <td><p>Either <code>"normal"</code> or <code>"t"</code>. If
<code>likelihood = "t"</code>, then the user may provide the degrees
of freedom via <code>degrees_freedom</code>.</p></td>
    </tr>
    <tr>
      <th>limmashrink</th>
      <td><p>A logical. Should we apply hierarchical
shrinkage to the variances (<code>TRUE</code>) or not (<code>FALSE</code>)?
If <code>degrees_freedom = NULL</code> and <code>limmashrink = TRUE</code>
and <code>likelihood = "t"</code>, then we'll also use the limma
returned degrees of freedom.</p></td>
    </tr>
    <tr>
      <th>degrees_freedom</th>
      <td><p>if <code>likelihood = "t"</code>, then this is the
user-defined degrees of freedom for that distribution. If
<code>degrees_freedom</code> is <code>NULL</code> then the degrees of
freedom will be the sample size minus the number of covariates
minus <code>k</code>.</p></td>
    </tr>
    <tr>
      <th>include_intercept</th>
      <td><p>A logical. If <code>TRUE</code>, then it will
check <code>X</code> to see if it has an intercept term. If not, then
it will add an intercept term. If <code>FALSE</code>, then <code>X</code>
will be unchanged.</p></td>
    </tr>
    <tr>
      <th>gls</th>
      <td><p>A logical. Should we estimate the part of the
confounders associated with the nuisance parameters with gls
(<code>TRUE</code>) or with ols (<code>FALSE</code>).</p></td>
    </tr>
    <tr>
      <th>fa_func</th>
      <td><p>A factor analysis function. It must take parameters:
<code>Y</code> a data matrix of numerics, <code>r</code> a positive integer
for the rank, and <code>vr</code> a positive integer for the number
of the first rows that have a different variance than the last
rows. It must return: <code>alpha</code> a matrix for the factor
loadings, <code>Z</code> a matrix for the factors, <code>sig_diag</code> a
vector of the column-wise variances, and <code>lambda</code> a
numeric for the variance inflation of the first <code>vr</code> rows
of <code>Y</code>. The default function is <code><a href='pca_2step.html'>pca_2step</a></code>,
which is the main difference between RUV2 and this version.</p></td>
    </tr>
    <tr>
      <th>fa_args</th>
      <td><p>A list. Additional arguments you want to pass to
fa_func.</p></td>
    </tr>
    <tr>
      <th>use_factor</th>
      <td><p>A logical. Should we use the estimates of
<code>alpha</code> and <code>sig_diag</code> from the factor analysis
(<code>TRUE</code>), or re-estimate these using OLS as RUV2 does it
(<code>FALSE</code>)? Right now it's probably a bad idea to have the
settings <code>use_factor = TRUE, fa_limmashrink = TRUE,
limmashrink = TRUE</code> since then the variance estimates of the
control genes are being shrunk twice.</p></td>
    </tr>
    <tr>
      <th>force_check</th>
      <td><p>A logical. Are you REALLY sure you want to use
another fa_func (<code>FALSE</code>) or should I ask you again
(<code>TRUE</code>)?</p></td>
    </tr>
    <tr>
      <th>fa_limmashrink</th>
      <td><p>A logical. Should we shrink the variances
during the factor analysis step (<code>TRUE</code>) or not
(<code>FALSE</code>)?</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list whose elements are:</p>
<p><code>betahat</code> A matrix of numerics. The ordinary least
    squares estimates of the coefficients of the covariate of
    interest WHEN YOU ALSO INCLUDE THE ESTIMATES OF THE UNOBSERVED
    CONFOUNDERS.</p>
<p><code>sebetahat</code> A matrix of positive numerics. This is the
    post-inflation adjusted standard errors for <code>ruv$betahat</code>.</p>
<p><code>tstats</code> A vector of numerics. The t-statistics for
    testing against the null hypothesis of the coefficient of the
    covariate of interest being zero. This is after estimating the
    variance inflation parameter.</p>
<p><code>pvalues</code> A vector of numerics. The p-values of said test
    above.</p>
<p><code>alphahat</code> A matrix of numerics. The estimates of the
    coefficients of the hidden confounders. Only identified up to a
    rotation on the rowspace.</p>
<p><code>Zhat</code> A matrix of numerics. The estimates of the
    confounding variables. Only identified up to a rotation on the
    columnspace.</p>
<p><code>sigma2</code> A vector of positive numerics. The estimates of
    the variances PRIOR to inflation.</p>
<p><code>sigma2_adjusted</code> A vector of positive numerics. The
    estimates of the variances AFTER to inflation. This is equal to
    <code>sigma2 * multiplier</code>.</p>
<p><code>multiplier</code> A numeric. The estimated variance inflation
    parameter.</p>
<p><code>mult_matrix</code> A matrix of numerics. Equal to
    <code><a href='https://rdrr.io/r/base/solve.html'>solve(t(cbind(X, Zhat)) %*% cbind(X, Zhat))</a></code>. One
    multiplies <code>sigma2</code> or <code>simga2_adjused</code> by the
    diagonal elements of <code>mult_matrix</code> to get the standard
    errors of <code>betahat</code>.</p>
<p><code>degrees_freedom</code> The degrees of freedom of the t-
    statistics.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>See <code><a href='vruv4.html'>vruv4</a></code> for a description of the model.</p>
<p>You can provide your own factor analysis, but it must include an
estimate for the variance inflation parameter. This turns out to be
pretty hard. The way I do it now seems to work OK.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    
<ul>
<li><p>Buja, A. and Eyuboglu, N., 1992. "Remarks on parallel analysis." <em>Multivariate behavioral research</em>, 27(4), pp.509-540. doi: <a href='https://doi.org/10.1207/s15327906mbr2704_2'>10.1207/s15327906mbr2704_2</a></p></li>
<li><p>Gagnon-Bartsch, J., Laurent Jacob, and Terence P. Speed, 2013. "Removing unwanted variation from high dimensional data with negative controls." Berkeley: Department of Statistics. University of California. <a href='https://statistics.berkeley.edu/tech-reports/820'>https://statistics.berkeley.edu/tech-reports/820</a></p></li>
<li><p>Gerard, David, and Matthew Stephens. 2021. "Unifying and Generalizing Methods for Removing Unwanted Variation Based on Negative Controls." <em>Statistica Sinica</em>, 31(3), 1145-1166. doi: <a href='https://doi.org/10.5705/ss.202018.0345'>10.5705/ss.202018.0345</a></p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='pca_2step.html'>pca_2step</a></code> for the special factor analysis
    that results in variance inflation in RUV2.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>David Gerard</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Gerard, Peter Carbonetto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


